### ISSUE TITLE
HTTP template

### ISSUE THREAD

**Issue by mumoshu** [argoproj/argo-workflows#795]
**This is a feature request.**

Hi, thanks for maintaining a great project!

## Context/Problem

I am the primary maintainer of [kube-aws](https://github.com/kubernetes-incubator/kube-aws) and also one of maintainers of [Brigade](https://github.com/Azure/brigade).

I long wanted a K8s-native durable workflow engine to reliably run my brigade scripts for various operations on my kube-aws clusters.

Why the use-case matters to me and Argo?
Brigade is often confused as a workflow engine, but it is rather an scripting framework, like bash. There's no durability in brigade's workflow [written in js](https://brigade.sh/), and no reliability unless you write some vanilla js to explicitly do so.

In regard to reliability and durability, Argo is a relatively full-blown workflow engine, supported by step-level retries and manual resubmit of permanently failed workflows

That being said, what I'm investigating is to integrate Brigade with Argo somehow to leverage good parts from the both projects.

## Proposal

My current idea is that implementing a "http" template type in Argo, so that Argo step is able to delegate the body of the task via HTTP, similarly to what Argo do to delegate the body to K8S via K8S API.

Why don't just use existing template to achieve that? Efficiency. I don't want to spin up a docker container just for making a single http request.

Would this makes sense to be covered by Argo?
If so, would you mind if I send a PR for this?

Any comments are welcomed!
Thanks

**Comment by mumoshu** [comment]
Or e.g. a shellout-based plugin system would work.

It will require you to bake a custom command/script into a custom Argo controller docker image, which is then used by the controller to serve a custom template type.

But I'm not sure we can decide to introduce or design such a general plugin system at this point.
I don't care about breaking changes in an edge feature like this, but still..

WDYT?

**Comment by jessesuen** [comment]
Hi @mumoshu, very interesting proposal. I agree there should be a convenient syntax to offload tasks to external actors to support non-container-based workloads. And HTTP seems like the most natural way to do this. Please correct me if misunderstand, but it seems like what you are proposing is akin to webhook functionality from a workflow.

I believe that you are proposing that the HTTP calls happen in the controller itself. But I have some concerns with performing the call in the controller which I will outline below:

* I assume these would be blocking HTTP calls which return once the work is complete. First, we need to avoid introducing any long, synchronous blocking calls in the controller's operator logic. The current design is that there are small number of workers (currently hard-wired to 8) which rely on millisecond operation times in order for these workers to handle many concurrent in-flight workflows. If a blocking call is introduced, the workers will easily become exhausted, effectively shutting down the controller. So to support this, the controller will need to offload the requests to another pool of HTTP workers in order to keep the operate time low. It then becomes trickier to (1) to ensure that the HTTP call are made only once -- such as in the case of an informer resync event and (2) have proper cancellation support (e.g. the workflow is deleted, or eventual [kill feature](https://github.com/argoproj/argo/issues/527)). Both of these are doable, but it just complicates the logic.

* There is a security concern about performing the call in the controller itself. Keep in mind for this feature, we would also need to support authenticated HTTP requests. If this is to be done in the controller, it means the controller itself will need direct access to secrets, something it currently does not need access to. Not only will the RBAC privileges for the workflow-controller's service account need to be escalated, but the users will required to store their secrets in the same namespace as the controller (kube-system by default). The current security model is that secrets are accessed in the namespace that the workflow is executing in, which allows the controller to defer responsibility of secret access, while also enabling multi-tenancy of workflows. If secrets need to be stored where the controller can access it, it breaks this security model.

In summary, I do like the idea of a new HTTP template type and syntax to make this convenient for users, but I do think it needs to be done as a container for the reasons described above. Spinning up a container for a single HTTP request might not be ideal, but we can mitigate this by setting the pod resource limits/requests to a bare minimum, and baking the implementation into the argoexec container image so that there won't be latency related to image pull (the assumption being that argoexec will have already been pulled).

**Comment by mumoshu** [comment]
@jessesuen Thank you very much for your detailed response!

I read your write up twice and I now believe that I understand both (1) the current architecture of argo with limited nums of controller and worker is preferable and (2)the current security model of argo controller does limit the attack surface of argo.

Good points!

Now, following your comments:

>  I do like the idea of a new HTTP template type and syntax to make this convenient for users

Yes, I do believe the syntax would help.

>  I do think it needs to be done as a container for the reasons described above

Can't agree more,

> Spinning up a container for a single HTTP request might not be ideal, but we can mitigate this by setting the pod resource limits/requests to a bare minimum

Ack,

> baking the implementation into the argoexec container image so that there won't be latency related to image pull (the assumption being that argoexec will have already been pulled).

I believe the implementation would be easy and useful, but also I'm a bit hesitant to do so because potential to bloat up `argoexec`'s code-base.

Probably we want argoexec to sized as small as possible now and in the future, right?

I believe many of argo's "step"(can i call it that way?) templates currently are just thin wrappers around `createWorkflowPod(nodeName, mainCtr, tmpl)` according to what I see in implementations of:

- [executeScript](https://github.com/argoproj/argo/blob/a45bf1b7558b3eb60ec65d02c166c306e7797a79/workflow/controller/operator.go#L1154)
- [executeResource](https://github.com/argoproj/argo/blob/a45bf1b7558b3eb60ec65d02c166c306e7797a79/workflow/controller/operator.go#L1348)
- [exexcuteContainer](https://github.com/argoproj/argo/blob/a45bf1b7558b3eb60ec65d02c166c306e7797a79/workflow/controller/operator.go#L1088)

I especially like the implementations of `executeScript` and `executeContainer` for its simplicity, usability benefits, while not bloating up argoexec. Likewise, I wish if we could just extend argo's workflow definition without touching argoexec.

**Comment by mumoshu** [comment]
Sorry for answering your question in reverse order!

> proposing is akin to webhook functionality from a workflow.

Probably I'd say no, because what I expect from a "webhook" is asynchronous transaction call.

FWIW, what I had in my mind at the time of proposing this is a synchronous one, a request-reply-style RPC call over HTTP(almost certainly a REST API call).

And given the discussion above, I now understand this kind of blocking-calls doesn't fit nicely within the current architecture of Argo with low concurrency and parallelism in controller and workers.

**Comment by mumoshu** [comment]
@jessesuen Just opened up #799. In nutshell, I realized that we can just utilize templates to concisely write my workflow.

Given [the security and performance concerns](https://github.com/argoproj/argo/issues/795#issuecomment-373661969), I believe what I outlined in #799 is the best way to achieve it implementation-wise.

Still, just wrapping a common pattern of `template` within an out-of-box template type like `http` could be possible on top. But I don't see a huge benefit from doing so. Therefore, I'm suggesting only doc enchancement in #799. WDYT?

**Comment by jessesuen** [comment]
Hi @mumoshu. Sorry for the late reply (was on leave for past few weeks). I think if your requirements can be satisfied with a pattern instead of a new template type, we can go with that. If we see repeated requests for something like this we can re-evaluate that in the future. We appreciate you sharing your ideas & contributions!

**Comment by wanghong230** [comment]
@mumoshu The proposal sounds like Netflix Conductor to me.

**Comment by varunbpatil** [comment]
Being able to make HTTP requests without launching a new pod for each request would be an awesome feature to have for orchestrating workflows in FaaS (ex: kubeless).

**Comment by Bobgy** [comment]
Quoted from https://github.com/argoproj/argo/issues/799

> What is the performance(latency/throughput) implication by implementing "macros" this way?

> Back to #795, one of my expectations was that executing http templates directly within Argo controller would have a potential for achieving:

> higher throughput
lower latency
lower cpu/memory consumption
than running it this way(=nested workflow processing and at least one k8s pod creation per macro call).

These are still valid reasons a http template is preferable than pod templates.

**Comment by alexec** [comment]
We don't use ❤️  for prioritising, we use 👍 . Please thumbs up rather than heart.

**Comment by alexec** [comment]
@Bobgy I'm trying to ascertain if we should implement a 1st-class HTTP template, or a more general RPC plugin template.

**Comment by Bobgy** [comment]
Hi @alexec 

Thank a lot for keep moving this forward!

1st party http template is a perfect match for my use-case. Also, I think its interface is simpler than a generic plugin (so I think it would be easier to maintain too for argo).

However, I understand your concerns raised in the issue. They are not really a problem for my specific use-case. I'll leave it to you to decide if adding this template is overall better than plugins or not.

If you prefer supporting plugins, there will be a bit more work for us to build the plugin (acceptable), but that also makes our users rely on our configured distribution of argo. There might be people running the same cluster with both argo and KFP, I think that would in general create more segmentation.

These are my personal opinions on this issue.

**Comment by alexec** [comment]
I'm not sure. I'm currently leaning towards just making a built-in template type. Easier than writing a whole plugin system, to write just one common plugin.

**Comment by alexec** [comment]
See #5025

**Comment by alexec** [comment]
Notes from chat with KFP:

* A v1 should be sync. Async could be v2.
* Move useful that lifecycle hooks (#4355 )
* More useful than plugin template (#5025 )
